lib_eio_linux/eio_linux.ml

```diff
@@ -1162,11 +1162,90 @@ let net = object
   method getnameinfo = Eio_unix.getnameinfo
 end
 
+let rec wait_for_process flags pid = 
+  try Unix.waitpid flags pid with
+  | Unix.Unix_error (Unix.EINTR, _, _) -> wait_for_process flags pid
+
+external sig_to_host_sig : int -> int = "caml_signal_to_posix_signal"
+
+let pid_to_process close pid = object
+  inherit Eio.Process.t
+  method pid = pid
+  val mutable needs_close = true
+  method needs_close = needs_close
+  method set_needs_close b = needs_close <- b
+  method status =
+    let p, r = Promise.create () in
+    let run () =
+      let v = wait_for_process [] pid in 
+      if needs_close then (close (); needs_close <- false);
+      match v with
+      | _, WEXITED i -> Promise.resolve r  (Eio.Process.Exited (sig_to_host_sig i))
+      | _, WSIGNALED i -> Promise.resolve r  (Eio.Process.Signaled (sig_to_host_sig i))
+      | _, WSTOPPED i -> Promise.resolve r  (Eio.Process.Stopped (sig_to_host_sig i))
+    in
+    Eio_unix.run_in_systhread run;
+    p
+  
+  method stop = 
+    Unix.kill pid Sys.sigkill
+end
+
+let get_fd_or_err flow =
+  match get_fd_opt flow with
+  | Some fd -> 
+    Unix.dup ~cloexec:false (FD.to_unix `Peek fd)
```

Comment:
Setting an FD to be not close-on-exec needs to be done in the child process (in C). Otherwise, another domain might fork a process at the same time and inherit it.